File: src/weaponsprocurement/gui/WimGuiDialogTracker.java
package weaponsprocurement.gui;

final class WimGuiDialogTracker<C, S> {
    private boolean open = false;
    private C pendingContext;
    private S pendingState;

    boolean isOpen() {
        return open;
    }

    void markOpen() {
        open = true;
    }

    void markClosed() {
        open = false;
    }

    void requestReopen(C context, S state) {
        pendingContext = context;
        pendingState = state;
        open = false;
    }

    boolean hasPending() {
        return !open && pendingState != null;
    }

    WimGuiPendingDialog<C, S> consumePending() {
        if (!hasPending()) {
            return null;
        }
        WimGuiPendingDialog<C, S> pending = new WimGuiPendingDialog<C, S>(pendingContext, pendingState);
        pendingContext = null;
        pendingState = null;
        return pending;
    }
}

File: src/main/kotlin/weaponsprocurement/ui/WimGuiDialogTracker.kt
package weaponsprocurement.ui

class WimGuiDialogTracker<C, S> {
    private var open = false
    private var pendingContext: C? = null
    private var pendingState: S? = null

    fun isOpen(): Boolean = open

    fun markOpen() {
        open = true
    }

    fun markClosed() {
        open = false
    }

    fun requestReopen(context: C, state: S) {
        pendingContext = context
        pendingState = state
        open = false
    }

    fun hasPending(): Boolean = !open && pendingState != null

    fun consumePending(): WimGuiPendingDialog<C?, S?>? {
        if (!hasPending()) return null
        val pending = WimGuiPendingDialog(pendingContext, pendingState)
        pendingContext = null
        pendingState = null
        return pending
    }
}
